File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    
    vector<vector<int>> ans;
    
    void solve(int i,vector<int>& candidates,int target,vector<int>& temp)
    {
        //3 base case
        if(target==0)
        {
            ans.push_back(temp);
            return ;
            }
        
        if(target<0)
            return ;
        if(i==candidates.size())//means we have reached out of  array
            return ;
        
        
        
        //first type of case case where we are not taking that element in recursion
        solve(i+1,candidates,target,temp);
        
        
        
        temp.push_back(candidates[i]);
        //second case where we are taking back the same element again
          solve(i,candidates,target-candidates[i],temp);
        temp.pop_back();
        
        
    }
    
    
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
       vector<int> temp;
        
        solve(0,candidates,target,temp);
        
        return ans;
        
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    @staticmethod
    def function(index,arr,target,temp,result):
        if index==len(arr):
            if target==0:
                result.append(temp[:])
            return
        
        
        
        
        # to pick the number at given index posn
        if (arr[index]<=target):
            temp.append(arr[index])
            Solution.function(index,arr,target-arr[index],temp,result)
            temp.pop()
            
        
        
        Solution.function(index+1,arr,target,temp,result)
        
  
        
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        temp=[]
        result=[]
        Solution.function(0,candidates,target,temp,result)
        return result
        